presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -348,14 +395,38 @@ public Type getType()
 
         public void processPage(GroupByIdBlock groupIds, Page page)
         {
-            if (step.isInputRaw()) {
+            if (!this.orderByChannels.isEmpty()) {
+                checkArgument(step == Step.SINGLE && pageIndex.isPresent(), "Order by over distributed aggregation is not supported");
+                pageIndex.ifPresent(value -> value.addPage(page));
+            }
+            else if (step.isInputRaw()) {
                 aggregation.addInput(groupIds, page);
             }
             else {
                 aggregation.addIntermediate(groupIds, page.getBlock(intermediateChannel));
             }
         }
 
+        public void processOrderBy()
```

Comment:
I don't have strong opinions for or against this. It is logically clearer to put all order related things into another class. But the logic is not complicated to cause confusion as it is now. And the new class, if added, will only be used here, private to an already private static class. So if you guys like the idea better, I can make the change. Otherwise keep it as is is less work for me. :P